# Introduction

Hopefully, this short tutorial can show you a lot of different commands that will help you gain the most insights into your dataset. 

In [9]:
# Import the pandas library for handling and analyzing data
import pandas as pd

# Import sys to access system-specific parameters and functions
import sys

# Import os to interact with the operating system, like file handling
import os

# Modify the system path to include the parent directory for module imports
sys.path.append(os.path.abspath(".."))

# Import the 'load_data_from_google_drive' function from the utils module inside the 'src' folder
from src.utils import load_data_from_google_drive


# Loading in Data

The first step in any ML problem is identifying what format your data is in, and then loading it into whatever framework you're using. For Kaggle compeitions, a lot of data can be found in CSV files, so that's the example we're going to use. 

We're going to be looking at a sports dataset that shows the results from NCAA basketball games from 1985 to 2016. This dataset is in a CSV file, and the function we're going to use to read in the file is called **pd.read_csv()**. This function returns a **dataframe** variable. The dataframe is the golden jewel data structure for Pandas. It is defined as "a two-dimensional size-mutable, potentially heterogeneous tabular data structure with labeled axes (rows and columns)".

Just think of it as a table for now. 

In [10]:
df = load_data_from_google_drive(url='https://drive.google.com/file/d/184JcLbSpArA_uq0DgAv2k892KChJVPHt/view?usp=share_link')

In [11]:
df

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0         1985      20   1228      81   1328      64    N      0
1         1985      25   1106      77   1354      70    H      0
2         1985      25   1112      63   1223      56    H      0
3         1985      25   1165      70   1432      54    H      0
4         1985      25   1192      86   1447      74    H      0
...        ...     ...    ...     ...    ...     ...  ...    ...
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

[145289 rows x 8 columns]

# The Basics

Now that we have our dataframe in our variable df, let's look at what it contains. We can use the function **head()** to see the first couple rows of the dataframe (or the function **tail()** to see the last few rows).

In [12]:
df.head()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0
4    1985      25   1192      86   1447      74    H      0

In [13]:
df.tail()

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
145284    2016     132   1114      70   1419      50    N      0
145285    2016     132   1163      72   1272      58    N      0
145286    2016     132   1246      82   1401      77    N      1
145287    2016     132   1277      66   1345      62    N      0
145288    2016     132   1386      87   1433      74    N      0

We can see the dimensions of the dataframe using the the **shape** attribute

In [14]:
df.shape

(145289, 8)

We can also extract all the column names as a list, by using the **columns** attribute and can extract the rows with the **index** attribute

In [15]:
df.columns.tolist()

['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'Wloc', 'Numot']

In order to get a better idea of the type of data that we are dealing with, we can call the **describe()** function to see statistics like mean, min, etc about each column of the dataset. 

In [16]:
df.describe()

Season         Daynum          Wteam         Wscore  \
count  145289.000000  145289.000000  145289.000000  145289.000000   
mean     2001.574834      75.223816    1286.720646      76.600321   
std         9.233342      33.287418     104.570275      12.173033   
min      1985.000000       0.000000    1101.000000      34.000000   
25%      1994.000000      47.000000    1198.000000      68.000000   
50%      2002.000000      78.000000    1284.000000      76.000000   
75%      2010.000000     103.000000    1379.000000      84.000000   
max      2016.000000     132.000000    1464.000000     186.000000   

               Lteam         Lscore          Numot  
count  145289.000000  145289.000000  145289.000000  
mean     1282.864064      64.497009       0.044387  
std       104.829234      11.380625       0.247819  
min      1101.000000      20.000000       0.000000  
25%      1191.000000      57.000000       0.000000  
50%      1280.000000      64.000000       0.000000  
75%      1375.000000      72.000000       0.000000  
max      1464.000000     150.000000       6.000000

Okay, so now let's looking at information that we want to extract from the dataframe. Let's say I wanted to know the max value of a certain column. The function **max()** will show you the maximum values of all columns

In [17]:
df.max()

Season    2016
Daynum     132
Wteam     1464
Wscore     186
Lteam     1464
Lscore     150
Wloc         N
Numot        6
dtype: object

Then, if you'd like to specifically get the max value for a particular column, you pass in the name of the column using the bracket indexing operator

In [18]:
df['Wscore'].max()

186

If you'd like to find the mean of the Losing teams' score. 

In [19]:
df['Lscore'].mean()

64.49700940883343

But what if that's not enough? Let's say we want to actually see the game(row) where this max score happened. We can call the **argmax()** function to identify the row index

In [20]:
df['Wscore'].argmax()

24970

One of the most useful functions that you can call on certain columns in a dataframe is the **value_counts()** function. It shows how many times each item appears in the column. This particular command shows the number of games in each season

In [21]:
df['Season'].value_counts()

Season
2016    5369
2014    5362
2015    5354
2013    5320
2010    5263
2012    5253
2009    5249
2011    5246
2008    5163
2007    5043
2006    4757
2005    4675
2003    4616
2004    4571
2002    4555
2000    4519
2001    4467
1999    4222
1998    4167
1997    4155
1992    4127
1991    4123
1996    4122
1995    4077
1994    4060
1990    4045
1989    4037
1993    3982
1988    3955
1987    3915
1986    3783
1985    3737
Name: count, dtype: int64

**Q**: How many unique seasons are there in the dataset? Use the nunique() function.

In [23]:
# Use the nunique() function to find the number of unique seasons in the dataset
# 'Season' column represents the season for each basketball game
unique_seasons = df['Season'].nunique()

# Print the number of unique seasons
print(f"The number of unique seasons in the dataset is: {unique_seasons}")



The number of unique seasons in the dataset is: 32


**Q**: Find the team with the most wins. Use the value_counts() function on the Wteam column.

In [24]:
# Write your code here
# Use value_counts() to find the number of wins for each team
team_win_counts = df['Wteam'].value_counts()

# Find the team with the most wins (first element in the result)
most_wins_team = team_win_counts.idxmax()

# Print the team ID with the most wins
print(f"The team with the most wins is team ID: {most_wins_team}")


The team with the most wins is team ID: 1181


# Acessing Values

Then, in order to get attributes about the game, we need to use the **iloc[]** function. Iloc is definitely one of the more important functions. The main idea is that you want to use it whenever you have the integer index of a certain row that you want to access. As per Pandas documentation, iloc is an "integer-location based indexing for selection by position."

In [25]:
df.iloc[[df['Wscore'].argmax()]]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
24970    1991      68   1258     186   1109     140    H      0

Let's take this a step further. Let's say you want to know the game with the highest scoring winning team (this is what we just calculated), but you then want to know how many points the losing team scored. 

In [26]:
df.iloc[[df['Wscore'].argmax()]]['Lscore']

24970    140
Name: Lscore, dtype: int64

When you see data displayed in the above format, you're dealing with a Pandas **Series** object, not a dataframe object.

In [27]:
type(df.iloc[[df['Wscore'].argmax()]]['Lscore'])

pandas.core.series.Series

In [28]:
type(df.iloc[[df['Wscore'].argmax()]])

pandas.core.frame.DataFrame

The following is a summary of the 3 data structures in Pandas (Haven't ever really used Panels yet)

![](DataStructures.png)

When you want to access values in a Series, you'll want to just treat the Series like a Python dictionary, so you'd access the value according to its key (which is normally an integer index)

In [29]:
df.iloc[[df['Wscore'].argmax()]]['Lscore'][24970]

140

The other really important function in Pandas is the **loc** function. Contrary to iloc, which is an integer based indexing, loc is a "Purely label-location based indexer for selection by label". Since all the games are ordered from 0 to 145288, iloc and loc are going to be pretty interchangable in this type of dataset

In [30]:
df.iloc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0

In [31]:
df.loc[:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot
0    1985      20   1228      81   1328      64    N      0
1    1985      25   1106      77   1354      70    H      0
2    1985      25   1112      63   1223      56    H      0
3    1985      25   1165      70   1432      54    H      0

Notice the slight difference in that iloc is exclusive of the second number, while loc is inclusive. 

Below is an example of how you can use loc to acheive the same task as we did previously with iloc

In [32]:
df.loc[df['Wscore'].argmax(), 'Lscore']

140

A faster version uses the **at()** function. At() is really useful wheneever you know the row label and the column label of the particular value that you want to get. 

In [33]:
df.at[df['Wscore'].argmax(), 'Lscore']

140

If you'd like to see more discussion on how loc and iloc are different, check out this great Stack Overflow post: http://stackoverflow.com/questions/31593201/pandas-iloc-vs-ix-vs-loc-explanation. Just remember that **iloc looks at position** and **loc looks at labels**. Loc becomes very important when your row labels aren't integers. 

# Sorting

Let's say that we want to sort the dataframe in increasing order for the scores of the losing team

In [ ]:
df.sort_values('Lscore').head()

**Q**: Make three dataframes that are sorted by season, winning team, and winning score respectively. Then, Using iloc, select the rows from index 100 to 200 and the columns for season, winning team, and winning score, respectively. 

In [34]:
# Write your code here
# Sort the DataFrame by Season
df_sorted_by_season = df.sort_values('Season')

# Sort the DataFrame by Winning Team
df_sorted_by_wteam = df.sort_values('Wteam')

# Sort the DataFrame by Winning Score
df_sorted_by_wscore = df.sort_values('Wscore')

# Now, use iloc to select rows from index 100 to 200 and the desired columns
# Select columns for Season, Winning Team, and Winning Score (which are at indices 0, 2, and 3)
selected_rows = df_sorted_by_season.iloc[100:201, [0, 2, 3]]

# Display the selected rows
print(selected_rows)


      Season  Wteam  Wscore
2525    1985   1406      68
2526    1985   1410      83
2527    1985   1423      58
2528    1985   1429      86
2529    1985   1440      50
...      ...    ...     ...
2447    1985   1212      63
2448    1985   1216      64
2449    1985   1233      77
2450    1985   1249      80
2425    1985   1449      72

[101 rows x 3 columns]


**Q**: From these three subsets you obtained above, find the season and winning team for the game with the highest winning score.

In [35]:
# Write your code here
# Find the row with the highest winning score in the sorted-by-wscore DataFrame
max_winning_score_index = df_sorted_by_wscore['Wscore'].argmax()

# Retrieve the corresponding season and winning team for that game
season_of_max_score = df_sorted_by_wscore.at[max_winning_score_index, 'Season']
winning_team_of_max_score = df_sorted_by_wscore.at[max_winning_score_index, 'Wteam']

# Display the results
print(f"Season with highest winning score: {season_of_max_score}")
print(f"Winning team with highest winning score: {winning_team_of_max_score}")


Season with highest winning score: 2016
Winning team with highest winning score: 1386


# Filtering Rows Conditionally

Now, let's say we want to find all of the rows that satisy a particular condition. For example, I want to find all of the games where the winning team scored more than 150 points. The idea behind this command is you want to access the column 'Wscore' of the dataframe df (df['Wscore']), find which entries are above 150 (df['Wscore'] > 150), and then returns only those specific rows in a dataframe format (df[df['Wscore'] > 150]).

In [ ]:
df[df['Wscore'] > 150]

This also works if you have multiple conditions. Let's say we want to find out when the winning team scores more than 150 points and when the losing team scores below 100. 

In [ ]:
df[(df['Wscore'] > 150) & (df['Lscore'] < 100)]

**Q**: Create a new column in the DataFrame called 'ScoreDifference' which is the absolute difference between the winning score and the losing score. Filter the DataFrame to only include games where the 'ScoreDifference' is greater than the average 'ScoreDifference' for all games.

In [37]:
# Write your code here
# Step 1: Create the 'ScoreDifference' column
df['ScoreDifference'] = abs(df['Wscore'] - df['Lscore'])

# Step 2: Calculate the average 'ScoreDifference' for all games
average_score_diff = df['ScoreDifference'].mean()

# Step 3: Filter the DataFrame to include games where 'ScoreDifference' is greater than the average
filtered_df = df[df['ScoreDifference'] > average_score_diff]

# Display the filtered DataFrame
print(filtered_df)


        Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  \
0         1985      20   1228      81   1328      64    N      0   
3         1985      25   1165      70   1432      54    H      0   
6         1985      25   1228      64   1226      44    N      0   
8         1985      25   1260      98   1133      80    H      0   
10        1985      25   1307     103   1288      71    H      0   
...        ...     ...    ...     ...    ...     ...  ...    ...   
145280    2016     131   1401      71   1261      38    N      0   
145282    2016     131   1433      76   1172      54    N      0   
145284    2016     132   1114      70   1419      50    N      0   
145285    2016     132   1163      72   1272      58    N      0   
145288    2016     132   1386      87   1433      74    N      0   

        ScoreDifference  
0                    17  
3                    16  
6                    20  
8                    18  
10                   32  
...                 ...  
1

**Q**: From this filtered DataFrame, find the season and teams involved in the game with the highest 'ScoreDifference'.

In [39]:
# Write your code here

# Step 1: Find the row with the maximum 'ScoreDifference' in the filtered DataFrame
max_score_diff_row = filtered_df.loc[filtered_df['ScoreDifference'].idxmax()]

# Step 2: Extract the season, winning team, and losing team for that row
season = max_score_diff_row['Season']
winning_team = max_score_diff_row['Wteam']
losing_team = max_score_diff_row['Lteam']

# Display the results
print(f"Season: {season}, Winning Team: {winning_team}, Losing Team: {losing_team}")



Season: 1996, Winning Team: 1409, Losing Team: 1341


# Grouping

Another important function in Pandas is **groupby()**. This is a function that allows you to group entries by certain attributes (e.g Grouping entries by Wteam number) and then perform operations on them. The following function groups all the entries (games) with the same Wteam number and finds the mean for each group. 

In [40]:
df.groupby('Wteam')['Wscore'].mean().head()

Wteam
1101    78.111111
1102    69.893204
1103    75.839768
1104    75.825944
1105    74.960894
Name: Wscore, dtype: float64

This next command groups all the games with the same Wteam number and finds where how many times that specific team won at home, on the road, or at a neutral site

In [41]:
df.groupby('Wteam')['Wloc'].value_counts().head(9)

Wteam  Wloc
1101   H        12
       A         3
       N         3
1102   H       204
       A        73
       N        32
1103   H       324
       A       153
       N        41
Name: count, dtype: int64

Each dataframe has a **values** attribute which is useful because it basically displays your dataframe in a numpy array style format

In [42]:
df.values

array([[1985, 20, 1228, ..., 'N', 0, 17],
       [1985, 25, 1106, ..., 'H', 0, 7],
       [1985, 25, 1112, ..., 'H', 0, 7],
       ...,
       [2016, 132, 1246, ..., 'N', 1, 5],
       [2016, 132, 1277, ..., 'N', 0, 4],
       [2016, 132, 1386, ..., 'N', 0, 13]], dtype=object)

Now, you can simply just access elements like you would in an array. 

In [43]:
df.values[0][0]

1985

**Q**: Group the DataFrame by season and find the average winning score for each season.

In [44]:
# Write your code here
# Group the DataFrame by 'Season' and calculate the average winning score for each season
average_winning_score_per_season = df.groupby('Season')['Wscore'].mean()

# Display the result
print(average_winning_score_per_season)


Season
1985    74.723040
1986    74.813640
1987    77.993870
1988    79.773704
1989    81.728511
1990    80.846477
1991    82.733932
1992    79.992004
1993    79.508790
1994    81.160345
1995    80.096149
1996    78.205483
1997    76.300842
1998    77.229902
1999    76.043581
2000    76.310467
2001    77.088650
2002    76.963557
2003    75.795061
2004    74.971122
2005    74.693690
2006    74.587555
2007    74.829467
2008    74.962425
2009    74.033721
2010    74.652480
2011    74.374190
2012    73.379973
2013    72.840789
2014    76.101268
2015    72.680799
2016    78.295958
Name: Wscore, dtype: float64


**Q**: Group the DataFrame by winning team and find the maximum winning score for each team across all seasons.

In [45]:
# Write your code here
# Group the DataFrame by 'Wteam' and calculate the maximum winning score for each team
max_winning_score_per_team = df.groupby('Wteam')['Wscore'].max()

# Display the result
print(max_winning_score_per_team)


Wteam
1101     95
1102    111
1103    109
1104    114
1105    114
       ... 
1460    136
1461    112
1462    125
1463    105
1464    115
Name: Wscore, Length: 364, dtype: int64


**Q**: Group the DataFrame by both season and winning team. Find the team with the highest average winning score for each season.

In [46]:
# Write your code here
# Group the DataFrame by both 'Season' and 'Wteam', and calculate the average winning score for each team in each season
average_winning_scores = df.groupby(['Season', 'Wteam'])['Wscore'].mean()

# For each season, find the team with the highest average winning score
highest_avg_winning_score_team = average_winning_scores.groupby('Season').idxmax()

# Display the result
print(highest_avg_winning_score_team)


Season
1985    (1985, 1328)
1986    (1986, 1109)
1987    (1987, 1380)
1988    (1988, 1258)
1989    (1989, 1258)
1990    (1990, 1258)
1991    (1991, 1380)
1992    (1992, 1380)
1993    (1993, 1380)
1994    (1994, 1380)
1995    (1995, 1206)
1996    (1996, 1206)
1997    (1997, 1254)
1998    (1998, 1395)
1999    (1999, 1317)
2000    (2000, 1395)
2001    (2001, 1395)
2002    (2002, 1242)
2003    (2003, 1395)
2004    (2004, 1194)
2005    (2005, 1311)
2006    (2006, 1190)
2007    (2007, 1440)
2008    (2008, 1440)
2009    (2009, 1377)
2010    (2010, 1440)
2011    (2011, 1449)
2012    (2012, 1370)
2013    (2013, 1377)
2014    (2014, 1322)
2015    (2015, 1322)
2016    (2016, 1146)
Name: Wscore, dtype: object


**Q**: Create a new DataFrame that counts the number of wins for each team in each season. This will involve grouping by both season and winning team, and then using the count() function.

In [47]:
# Write your code here
# Group by 'Season' and 'Wteam' and count the number of wins for each team in each season
wins_per_team_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='NumWins')

# Display the resulting DataFrame
print(wins_per_team_season)


       Season  Wteam  NumWins
0        1985   1102        5
1        1985   1103        9
2        1985   1104       21
3        1985   1106       10
4        1985   1108       19
...       ...    ...      ...
10167    2016   1460       20
10168    2016   1461       12
10169    2016   1462       27
10170    2016   1463       21
10171    2016   1464        9

[10172 rows x 3 columns]


**Q**: For each season, find the team with the most wins. This will involve creating a DataFrame similar to the one in task 5, and then using the idxmax() function for each season.

In [48]:
# Write your code here

# Step 1: Group by 'Season' and 'Wteam' and count the number of wins
wins_per_team_season = df.groupby(['Season', 'Wteam']).size().reset_index(name='NumWins')

# Step 2: For each season, find the team with the most wins using idxmax
most_wins_per_season = wins_per_team_season.loc[wins_per_team_season.groupby('Season')['NumWins'].idxmax()]

# Display the resulting DataFrame
print(most_wins_per_season)



      Season  Wteam  NumWins
217     1985   1385       27
342     1986   1181       32
818     1987   1424       33
861     1988   1112       31
1323    1989   1328       28
1551    1990   1247       29
1739    1991   1116       30
2088    1992   1181       28
2423    1993   1231       28
2665    1994   1163       27
2928    1995   1116       26
3357    1996   1269       31
3638    1997   1242       31
3943    1998   1242       33
4201    1999   1181       32
4709    2000   1409       29
4829    2001   1181       29
5121    2002   1153       30
5456    2003   1166       29
5981    2004   1390       29
6165    2005   1228       32
6452    2006   1181       30
6841    2007   1242       30
7209    2008   1272       33
7552    2009   1272       31
7870    2010   1242       32
8217    2011   1242       32
8565    2012   1246       32
8878    2013   1211       30
9462    2014   1455       33
9609    2015   1246       34
9955    2016   1242       29


**Q**: Group the DataFrame by losing team and find the average losing score for each team across all seasons. Compare this with the average winning score for each team from task 3. Are there teams that have a higher average losing score than winning score?

In [49]:
# Write your code here
# Step 1: Group by losing team ('Lteam') and calculate the average losing score for each team
avg_losing_scores = df.groupby('Lteam')['Lscore'].mean().reset_index(name='AvgLscore')

# Step 2: Using the avg_winning_scores DataFrame from Task 3
# avg_winning_scores should already be available from previous calculations
# If it wasn't explicitly calculated earlier, calculate it again here:
avg_winning_scores = df.groupby('Wteam')['Wscore'].mean().reset_index(name='AvgWscore')

# Step 3: Merge the average winning scores with the average losing scores
team_scores = pd.merge(avg_winning_scores, avg_losing_scores, left_on='Wteam', right_on='Lteam', how='outer')

# Step 4: Find teams where the average losing score is higher than the average winning score
teams_with_higher_losing_scores = team_scores[team_scores['AvgLscore'] > team_scores['AvgWscore']]

# Display the result
print(teams_with_higher_losing_scores)


Empty DataFrame
Columns: [Wteam, AvgWscore, Lteam, AvgLscore]
Index: []


# Dataframe Iteration

In order to iterate through dataframes, we can use the **iterrows()** function. Below is an example of what the first two rows look like. Each row in iterrows is a Series object

In [50]:
for index, row in df.iterrows():
    print(row)
    if index == 1:
        break

Season             1985
Daynum               20
Wteam              1228
Wscore               81
Lteam              1328
Lscore               64
Wloc                  N
Numot                 0
ScoreDifference      17
Name: 0, dtype: object
Season             1985
Daynum               25
Wteam              1106
Wscore               77
Lteam              1354
Lscore               70
Wloc                  H
Numot                 0
ScoreDifference       7
Name: 1, dtype: object


**Q**: Create a new column 'HighScoringGame' that is 'Yes' if the winning score is greater than 100 and 'No' otherwise. This will require iterating over the rows of the DataFrame and checking the value of the winning score for each row.

In [65]:
# Write your code here

# Create the 'HighScoringGame' column directly using a vectorized condition
df['HighScoringGame'] = df['Wscore'].apply(lambda x: 'Yes' if x > 100 else 'No')

# Display the first few rows to verify
print(df[['Season', 'Daynum', 'Wteam', 'Wscore', 'HighScoringGame']].head())



   Season  Daynum  Wteam  Wscore HighScoringGame
0    1985      20   1228      81              No
1    1985      25   1106      77              No
2    1985      25   1112      63              No
3    1985      25   1165      70              No
4    1985      25   1192      86              No


**Q**: Calculate the total number of games played by each team, whether they won or lost. This will require iterating over the rows of the DataFrame and updating a dictionary that keeps track of the number of games for each team.

In [56]:
# Write your code here
# Convert the dictionary to a DataFrame for better visualization
games_played_df = pd.DataFrame(list(games_played.items()), columns=['Team', 'GamesPlayed'])
games_played_df.sort_values(by='GamesPlayed', ascending=False, inplace=True)

# Display the DataFrame with the number of games played
print(games_played_df.head())




     Team  GamesPlayed
106  1181         1013
98   1314         1010
136  1246          999
61   1257          994
13   1242          993


**Q**: For each season, find the game with the highest score difference (winning score - losing score). This will require iterating over the rows of the DataFrame, keeping track of the highest score difference for each season, and updating it if a game with a higher score difference is found.

In [57]:
# Write your code here
# Initialize a dictionary to store the highest score difference for each season
highest_score_diff = {}

# Iterate over the rows of the DataFrame
for index, row in df.iterrows():
    # Calculate the score difference for the current game
    score_diff = row['Wscore'] - row['Lscore']
    
    # Check if the season already exists in the dictionary
    if row['Season'] not in highest_score_diff:
        # If not, initialize a dictionary entry for that season
        highest_score_diff[row['Season']] = {'max_diff': score_diff, 'game': row}
    else:
        # If the current game has a higher score difference, update the dictionary
        if score_diff > highest_score_diff[row['Season']]['max_diff']:
            highest_score_diff[row['Season']]['max_diff'] = score_diff
            highest_score_diff[row['Season']]['game'] = row

# Extract the game with the highest score difference for each season
highest_score_diff_games = {
    season: data['game'] for season, data in highest_score_diff.items()
}

# Convert the results to a DataFrame for better visualization
highest_score_diff_df = pd.DataFrame.from_dict(highest_score_diff_games, orient='index')

# Display the games with the highest score difference for each season
print(highest_score_diff_df[['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'ScoreDifference']])



      Season  Daynum  Wteam  Wscore  Lteam  Lscore  ScoreDifference
1985    1985      33   1361     128   1288      68               60
1986    1986      60   1314     129   1264      45               84
1987    1987      51   1155     112   1118      39               73
1988    1988      40   1328     152   1147      84               68
1989    1989      64   1242     115   1135      45               70
1990    1990      26   1181     130   1217      54               76
1991    1991      73   1163     115   1148      47               68
1992    1992      30   1116     128   1126      46               82
1993    1993      86   1328     146   1197      65               81
1994    1994      47   1228     121   1152      52               69
1995    1995      26   1246     124   1404      50               74
1996    1996      48   1409     141   1341      50               91
1997    1997      56   1278     114   1106      34               80
1998    1998      33   1395     153   1410      

Remember, iterating over a DataFrame should generally be avoided if a vectorized operation can be used instead, as vectorized operations are usually much faster. However, these tasks are designed to give practice with DataFrame iteration for cases where it might be necessary.

Vectorized Operation Example: Create a new column 'HighScoringGame' in the DataFrame using a vectorized operation. This column should contain 'Yes' if the winning score is greater than 100 and 'No' otherwise. Use the np.where function from the numpy library for this task.

In [66]:
import numpy as np
df['HighScoringGame'] = np.where(df['Wscore'] > 100, 'Yes', 'No')

**Q**: Vectorized Operation: Calculate the total number of games played by each team, whether they won or lost. Instead of iterating over the DataFrame, use the value_counts() function on the winning team and losing team columns separately, and then add the two Series together.

In [67]:
# Write your code here
# Get the count of games where the team was a winning team
wins = df['Wteam'].value_counts()

# Get the count of games where the team was a losing team
losses = df['Lteam'].value_counts()

# Add the two Series together to get the total games played by each team
total_games = wins.add(losses, fill_value=0)

# Display the total number of games played by each team
print(total_games)


1101     76
1102    840
1103    910
1104    975
1105    447
       ... 
1460    827
1461    914
1462    954
1463    838
1464    856
Name: count, Length: 364, dtype: int64


**Q**: For each season, find the game with the highest score difference (winning score - losing score). Instead of iterating over the DataFrame, create a new column 'ScoreDifference' using vectorized subtraction, then use the groupby() function and idxmax() function to find the game with the highest score difference for each season.

In [68]:
# Write your code here

# Step 1: Create a new column 'ScoreDifference' using vectorized subtraction
df['ScoreDifference'] = df['Wscore'] - df['Lscore']

# Step 2: Group by 'Season' and find the index of the game with the maximum score difference for each season
max_score_diff_idx = df.groupby('Season')['ScoreDifference'].idxmax()

# Step 3: Use the indices to select the rows with the highest score difference for each season
max_score_diff_games = df.loc[max_score_diff_idx]

# Display the result
print(max_score_diff_games[['Season', 'Daynum', 'Wteam', 'Wscore', 'Lteam', 'Lscore', 'ScoreDifference']])


        Season  Daynum  Wteam  Wscore  Lteam  Lscore  ScoreDifference
236       1985      33   1361     128   1288      68               60
4731      1986      60   1314     129   1264      45               84
8240      1987      51   1155     112   1118      39               73
12046     1988      40   1328     152   1147      84               68
16677     1989      64   1242     115   1135      45               70
19502     1990      26   1181     130   1217      54               76
25161     1991      73   1163     115   1148      47               68
27997     1992      30   1116     128   1126      46               82
33858     1993      86   1328     146   1197      65               81
36404     1994      47   1228     121   1152      52               69
39858     1995      26   1246     124   1404      50               74
44653     1996      48   1409     141   1341      50               91
49033     1997      56   1278     114   1106      34               80
52600     1998      

# Extracting Rows and Columns

The bracket indexing operator is one way to extract certain columns from a dataframe.

In [69]:
df[['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Notice that you can acheive the same result by using the loc function. Loc is a veryyyy versatile function that can help you in a lot of accessing and extracting tasks. 

In [70]:
df.loc[:, ['Wscore', 'Lscore']].head()

Wscore  Lscore
0      81      64
1      77      70
2      63      56
3      70      54
4      86      74

Note the difference is the return types when you use brackets and when you use double brackets. 

In [71]:
type(df['Wscore'])

pandas.core.series.Series

In [72]:
type(df[['Wscore']])

pandas.core.frame.DataFrame

You've seen before that you can access columns through df['col name']. You can access rows by using slicing operations. 

In [73]:
df[0:3]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

Here's an equivalent using iloc

In [74]:
df.iloc[0:3,:]

Season  Daynum  Wteam  Wscore  Lteam  Lscore Wloc  Numot  ScoreDifference  \
0    1985      20   1228      81   1328      64    N      0               17   
1    1985      25   1106      77   1354      70    H      0                7   
2    1985      25   1112      63   1223      56    H      0                7   

  HighScoringGame  
0              No  
1              No  
2              No

# Data Cleaning

One of the big jobs of doing well in Kaggle competitions is that of data cleaning. A lot of times, the CSV file you're given (especially like in the Titanic dataset), you'll have a lot of missing values in the dataset, which you have to identify. The following **isnull** function will figure out if there are any missing values in the dataframe, and will then sum up the total for each column. In this case, we have a pretty clean dataset.

In [75]:
df.isnull().sum()

Season             0
Daynum             0
Wteam              0
Wscore             0
Lteam              0
Lscore             0
Wloc               0
Numot              0
ScoreDifference    0
HighScoringGame    0
dtype: int64

If you do end up having missing values in your datasets, be sure to get familiar with these two functions. 
* **dropna()** - This function allows you to drop all(or some) of the rows that have missing values. 
* **fillna()** - This function allows you replace the rows that have missing values with the value that you pass in.

# Other Useful Functions

* **drop()** - This function removes the column or row that you pass in (You also have the specify the axis). 
* **agg()** - The aggregate function lets you compute summary statistics about each group
* **apply()** - Lets you apply a specific function to any/all elements in a Dataframe or Series
* **get_dummies()** - Helpful for turning categorical data into one hot vectors.
* **drop_duplicates()** - Lets you remove identical rows

# Lots of Other Great Resources

Pandas has been around for a while and there are a lot of other good resources if you're still interested on getting the most out of this library. 
* http://pandas.pydata.org/pandas-docs/stable/10min.html
* https://www.datacamp.com/community/tutorials/pandas-tutorial-dataframe-python
* http://www.gregreda.com/2013/10/26/intro-to-pandas-data-structures/
* https://www.dataquest.io/blog/pandas-python-tutorial/
* https://drive.google.com/file/d/0ByIrJAE4KMTtTUtiVExiUGVkRkE/view
* https://www.youtube.com/playlist?list=PL5-da3qGB5ICCsgW1MxlZ0Hq8LL5U3u9y